ci: Supply `--platform` argument to `docker` commands.

davidgumberg opened the pull request:
I ran into this issue when following the instructions in `ci/README.md` for running CI locally.

Newer versions of docker require a `--platform` argument when building from a platform-specific image that differs from the host platform, I'm not sure when this change took place, but trying to build any of the cross-platform CI images on Docker 27.5.0 fails in the following manner:
```console
$ # From ci/README.md
$ env -i HOME="$HOME" PATH="$PATH" USER="$USER" bash -c 'FILE_ENV="./ci/test/00_setup_env_arm.sh" ./ci/test_run_all.sh'
WARNING: The requested image's platform (linux/arm64/v8) does not match the detected host platform (linux/amd64/v4) and no specific platform was requested
Creating docker.io/arm64v8/debian:bookworm container to run in
+ docker build --file $BITCOIN_SRC/ci/test_imagefile --build-arg CI_IMAGE_NAME_TAG=docker.io/arm64v8/debian:bookworm --build-arg FILE_ENV=./ci/test/00_setup_env_arm.sh --label=bitcoin-ci-test --tag=ci_arm_linux $BITCOIN_SRC
[+] Building 0.6s (2/2) FINISHED                                                                                                                                                 docker:default
 => [internal] load build definition from test_imagefile                                                                                                                                   0.0s
 => => transferring dockerfile: 600B                                                                                                                                                       0.0s
 => WARN: InvalidDefaultArgInFrom: Default value for ARG ${CI_IMAGE_NAME_TAG} results in empty or invalid base image name (line 8)                                                         0.0s
 => ERROR [internal] load metadata for docker.io/arm64v8/debian:bookworm                                                                                                                   0.5s
------
 > [internal] load metadata for docker.io/arm64v8/debian:bookworm:
------

 1 warning found (use docker --debug to expand):
 - InvalidDefaultArgInFrom: Default value for ARG ${CI_IMAGE_NAME_TAG} results in empty or invalid base image name (line 8)
test_imagefile:8
--------------------
   6 |
   7 |     ARG CI_IMAGE_NAME_TAG
   8 | >>> FROM ${CI_IMAGE_NAME_TAG}
   9 |
  10 |     ARG FILE_ENV
--------------------
ERROR: failed to solve: docker.io/arm64v8/debian:bookworm: failed to resolve source metadata for docker.io/arm64v8/debian:bookworm: no match for platform in manifest: not found
```

This branch fixes this by setting the `--platform` argument of `docker build` and `docker run` with an environment variable `CI_IMAGE_PLATFORM` for each platform specific job, and `linux/{$cpuarch}` for any native jobs.

Thi

## Steps to reproduce
1. Install relevant dependencies, on Ubuntu:
```bash
sudo apt install bash docker.io python3 qemu-user-static
```
2. Run one of the platform-specific CI images, e.g.:
```bash
env -i HOME="$HOME" PATH="$PATH" USER="$USER" bash -c 'FILE_ENV="./ci/test/00_setup_env_arm.sh" ./ci/test_run_all.sh'
```

davidgumberg committed:
ci: Supply `platform` argument to docker commands.

Later versions of docker require a `--platform` argument when building
and running from a platform-specific image that differs from the host
platform.

maflcko reviewed (approved):
lgtm, but I left some feedback

davidgumberg commented on ci/test/00_setup_env_s390x.sh:
```diff
@@ -10,6 +10,7 @@ export HOST=s390x-linux-gnu
 export PACKAGES="python3-zmq"
 export CONTAINER_NAME=ci_s390x
 export CI_IMAGE_NAME_TAG="docker.io/s390x/ubuntu:24.04"
+export DOCKER_DEFAULT_PLATFORM="linux/s390x"
```
> Does this env var work for podman as well?

The "podman" cli frontend doesn't use this environment variable[^1], but I tested locally (on Fedora 40) that this PR works for the build phase when using docker cli to interact with podman container host:

```console
$ sudo dnf install podman-machine
$ podman machine init
$ podman machine start
Starting machine "podman-machine-default"
API forwarding listening on: /run/user/1000/podman/podman-machine-default-api.sock
$ env -i HOME="$HOME" PATH="$PATH" USER="$USER" DOCKER_HOST='unix:///run/user/1000/podman/podman-machine-default-api.sock' bash -c 'FILE_ENV="./ci/test/00_setup_env_arm.sh" ./ci/test_run_all.sh'
WARNING: No output specified with docker-container driver. Build result will only remain in the build cache. To push result image into registry use --push or to load image into docker use --load
```
but as the warning hints will later fail when the `docker run` command is reached:
```console 
++ docker run --cap-add LINUX_IMMUTABLE --rm --interactive --detach --tty --mount [etc...] --env-file /tmp/env-{USER}-ci_arm_linux --name ci_arm_linux --network ci-ip6net ci_arm_linux
Unable to find image 'ci_arm_linux:latest' locally
$ cat /tmp/env-{USER}-ci_arm_linux | grep "DOCKER_HOST"
DOCKER_HOST=unix:///run/user/1000/podman/podman-machine-default-api.sock
```

but this can be fixed with:

```diff
--- a/ci/test/02_run_container.sh
+++ b/ci/test/02_run_container.sh
@@ -34,6 +34,7 @@ if [ -z "$DANGER_RUN_CI_ON_HOST" ]; then
       --label="${CI_IMAGE_LABEL}" \
       --tag="${CONTAINER_NAME}" \
+      --load \
```

-----
> Alternatively, with the added benefit of possibly supporting macos (see [#31344 (comment)](https://github.com/bitcoin/bitcoin/issues/31344#issuecomment-2493612110) (expand "docker setup")):
> 
> ```shell
> CI_IMAGE_PLATFORM="--os linux"  # native arch (default)
> ```

podman allows splitting up `--platform` into `--arch` and `--os` but docker doesn't :

```console
$ docker build --os linux
unknown flag: --os
$ docker build --arch s390x
unknown flag: --arch
```

---

I will try to troubleshoot what works best on an arm macos machine later today, but I guess the approach you are suggesting may work better if we ever want to support using `podman` for `build` and `run`.

[^1]: https://github.com/containers/podman/issues/11668

davidgumberg replied:
How to best make this compatible with podman and/or macos remains an open question, but I've switched to using `CI_IMAGE_PLATFORM` that gets passed to `docker build` and `docker run` as you suggested

davidgumberg replied:
Marking this as resolved since my latest push tries to set the right native "--platform=linux/{arch}", falling back to "--platform=linux" which is also valid according to docker's documentation although I haven't tested this yet.

davidgumberg commented on ci/test/00_setup_env_s390x.sh:
```diff
@@ -10,6 +10,7 @@ export HOST=s390x-linux-gnu
 export PACKAGES="python3-zmq"
 export CONTAINER_NAME=ci_s390x
 export CI_IMAGE_NAME_TAG="docker.io/s390x/ubuntu:24.04"
```
Thanks, done.

davidgumberg commented:
Rebased to address feedback, removed redundant arch specification from image names and switched to using `CI_IMAGE_PLATFORM="--platform os/arch"` that gets passed to `docker build` and `docker run` instead of using the `DOCKER_DEFAULT_PLATFORM` environment variable.

maflcko reviewed (commented):
lgtm, but you'll have to undo the last force-push

davidgumberg commented on ci/test/00_setup_env.sh:
```diff
@@ -68,3 +68,6 @@ export CI_BASE_PACKAGES=${CI_BASE_PACKAGES:-build-essential pkg-config curl ca-c
 export CI_RETRY_EXE=${CI_RETRY_EXE:-"retry --"}
+
+# The --platform argument used with `docker build` and `docker run`.
+export CI_IMAGE_PLATFORM=${CI_IMAGE_PLATFORM} # native platform by default
```
Thanks, fixed with a different approach.

davidgumberg commented on ci/test/00_setup_env.sh:
```diff
@@ -68,3 +68,16 @@ export CI_BASE_PACKAGES=${CI_BASE_PACKAGES:-build-essential pkg-config curl ca-c
 export DIR_QA_ASSETS=${DIR_QA_ASSETS:-${BASE_SCRATCH_DIR}/qa-assets}
 export CI_RETRY_EXE=${CI_RETRY_EXE:-"retry --"}
+
+# Try to get he host architecture in GOARCH format.
```
typo in this comment, I'll fix if I retouch

davidgumberg commented:
Rebased to incorporate feedback, now try to set a good default `--platform linux/$arch`  which should be an improvement for running CI on macOS.

This solution also avoids an issue I ran into after making the image tags less specific, which is that if no `--platform` argument is set, docker will try to use any image that it has cached before fetching a native image.

maflcko reviewed (approved):
lgtm ACK e1c7c29727faf2d0a5c6fa01505418aac26958fe

I think some code can be removed to be more streamlined, but not a blocker.

davidgumberg commented on ci/test/00_setup_env.sh:
```diff
@@ -68,3 +68,16 @@ export CI_BASE_PACKAGES=${CI_BASE_PACKAGES:-build-essential pkg-config curl ca-c
+    NATIVE_PLATFORM="linux"
+fi
+# The --platform argument used with `docker build` and `docker run`.
+export CI_IMAGE_PLATFORM=${CI_IMAGE_PLATFORM:-$NATIVE_PLATFORM} # native platform by default
```
The reason I did this was because of the caching issue I described above which can be reproduced as follows with a clean docker cache:

```console
$ # clean docker cache on fedora 40
$ systemctl restart docker 
$ docker system prune -a

$ docker run --platform=s390x docker.io/ubuntu:24.04 bash -c 'uname -m'
Unable to find image 'ubuntu:24.04' locally
24.04: Pulling from library/ubuntu
755503a8fb36: Pull complete
Digest: sha256:80dd3c3b9c6cecb9f1667e9290b3bc61b78c2678c02cbdae5f0fea92cc6734ab
Status: Downloaded newer image for ubuntu:24.04
s390x
$ docker run docker.io/ubuntu:24.04 bash -c 'uname -m' # no platform argument specified
WARNING: The requested image's platform (linux/s390x) does not match the detected host platform (linux/amd64/v4) and no specific platform was requested
s390x
```

but it seems docker infers the arch if an OS is specified, and doesn't just fall back on a cache hit for that OS:

```console
$ docker run --platform=linux docker.io/ubuntu:24.04 bash -c 'uname -m'
Unable to find image 'ubuntu:24.04' locally
24.04: Pulling from library/ubuntu
de44b265507a: Pull complete
Digest: sha256:80dd3c3b9c6cecb9f1667e9290b3bc61b78c2678c02cbdae5f0fea92cc6734ab
Status: Downloaded newer image for ubuntu:24.04
x86_64
```

davidgumberg commented:
rebased to address merge conflicts and dropped unnecessary logic for getting the host machine arch.

hebasto reviewed (commented):
Tested 6e29de21010fc5213176a6ba29f754ca72612ea0 on Ubuntu 24.10:
```
$ docker --version
Docker version 27.1.1, build 6312585
$ env -i HOME="$HOME" PATH="$PATH" USER="$USER" bash -c 'FILE_ENV="./ci/test/00_setup_env_arm.sh" ./ci/test_run_all.sh'
++++ dirname ./ci/test/00_setup_env.sh
+++ cd ./ci/test/../../
+++ pwd
++ BASE_READ_ONLY_DIR=/home/hebasto/git/bitcoin
++ export BASE_READ_ONLY_DIR
++ export BASE_ROOT_DIR=/ci_container_base
++ BASE_ROOT_DIR=/ci_container_base
++ export DEPENDS_DIR=/ci_container_base/depends
++ DEPENDS_DIR=/ci_container_base/depends
++ export BASE_SCRATCH_DIR=/ci_container_base/ci/scratch
++ BASE_SCRATCH_DIR=/ci_container_base/ci/scratch
++ export BINS_SCRATCH_DIR=/ci_container_base/ci/scratch/bins/
++ BINS_SCRATCH_DIR=/ci_container_base/ci/scratch/bins/
++ echo 'Setting specific values in env'
Setting specific values in env
++ '[' -n ./ci/test/00_setup_env_arm.sh ']'
++ set -o errexit
++ source ./ci/test/00_setup_env_arm.sh
+++ export LC_ALL=C.UTF-8
+++ LC_ALL=C.UTF-8
+++ export HOST=arm-linux-gnueabihf
+++ HOST=arm-linux-gnueabihf
+++ export DPKG_ADD_ARCH=armhf
+++ DPKG_ADD_ARCH=armhf
+++ export 'PACKAGES=python3-zmq g++-arm-linux-gnueabihf busybox libc6:armhf libstdc++6:armhf libfontconfig1:armhf libxcb1:armhf'
+++ PACKAGES='python3-zmq g++-arm-linux-gnueabihf busybox libc6:armhf libstdc++6:armhf libfontconfig1:armhf libxcb1:armhf'
+++ export CONTAINER_NAME=ci_arm_linux
+++ CONTAINER_NAME=ci_arm_linux
+++ export CI_IMAGE_NAME_TAG=docker.io/debian:bookworm
+++ CI_IMAGE_NAME_TAG=docker.io/debian:bookworm
+++ export CI_IMAGE_PLATFORM=linux/arm64
+++ CI_IMAGE_PLATFORM=linux/arm64
+++ export USE_BUSY_BOX=true
+++ USE_BUSY_BOX=true
+++ export RUN_UNIT_TESTS=true
+++ RUN_UNIT_TESTS=true
+++ export RUN_FUNCTIONAL_TESTS=false
+++ RUN_FUNCTIONAL_TESTS=false
+++ export GOAL=install
+++ GOAL=install
+++ export 'BITCOIN_CONFIG=-DREDUCE_EXPORTS=ON -DCMAKE_CXX_FLAGS='\''-Wno-psabi -Wno-error=maybe-uninitialized'\'''
+++ BITCOIN_CONFIG='-DREDUCE_EXPORTS=ON -DCMAKE_CXX_FLAGS='\''-Wno-psabi -Wno-error=maybe-uninitialized'\'''
++ echo 'Fallback to default values in env (if not yet set)'
Fallback to default values in env (if not yet set)
++ export MAKEJOBS=-j4
++ MAKEJOBS=-j4
++ export USE_BUSY_BOX=true
++ USE_BUSY_BOX=true
++ export RUN_UNIT_TESTS=true
++ RUN_UNIT_TESTS=true
++ export RUN_FUNCTIONAL_TESTS=false
++ RUN_FUNCTIONAL_TESTS=false
++ export RUN_TIDY=false
++ RUN_TIDY=false
++ export TEST_RUNNER_TIMEOUT_FACTOR=40
++ TEST_RUNNER_TIMEOUT_FACTOR=40
++ export RUN_FUZZ_TESTS=false
++ RUN_FUZZ_TESTS=false
++ export BOOST_TEST_RANDOM=1
++ BOOST_TEST_RANDOM=1
++ export DEBIAN_FRONTEND=noninteractive
++ DEBIAN_FRONTEND=noninteractive
++ export CCACHE_MAXSIZE=500M
++ CCACHE_MAXSIZE=500M
++ export CCACHE_TEMPDIR=/tmp/.ccache-temp
++ CCACHE_TEMPDIR=/tmp/.ccache-temp
++ export CCACHE_COMPRESS=1
++ CCACHE_COMPRESS=1
++ export CCACHE_DIR=/ci_container_base/ci/scratch/ccache
++ CCACHE_DIR=/ci_container_base/ci/scratch/ccache
++ export BASE_OUTDIR=/ci_container_base/ci/scratch/out
++ BASE_OUTDIR=/ci_container_base/ci/scratch/out
++ export PREVIOUS_RELEASES_DIR=/ci_container_base/prev_releases
++ PREVIOUS_RELEASES_DIR=/ci_container_base/prev_releases
++ export 'CI_BASE_PACKAGES=build-essential pkgconf curl ca-certificates ccache python3 rsync git procps bison e2fsprogs cmake'
++ CI_BASE_PACKAGES='build-essential pkgconf curl ca-certificates ccache python3 rsync git procps bison e2fsprogs cmake'
++ export GOAL=install
++ GOAL=install
++ export DIR_QA_ASSETS=/ci_container_base/ci/scratch/qa-assets
++ DIR_QA_ASSETS=/ci_container_base/ci/scratch/qa-assets
++ export 'CI_RETRY_EXE=retry --'
++ CI_RETRY_EXE='retry --'
++ export CI_IMAGE_PLATFORM=linux/arm64
++ CI_IMAGE_PLATFORM=linux/arm64
+ set -o errexit
+ ./ci/test/02_run_container.sh
+ '[' -z '' ']'
+ python3 -c 'import os; [print(f"{key}={value}") for key, value in os.environ.items() if "\n" not in value and "HOME" != key and "PATH" != key and "USER" != key]'
+ tee /tmp/env-hebasto-ci_arm_linux
CCACHE_TEMPDIR=/tmp/.ccache-temp
FILE_ENV=./ci/test/00_setup_env_arm.sh
CCACHE_MAXSIZE=500M
BASE_OUTDIR=/ci_container_base/ci/scratch/out
RUN_UNIT_TESTS=true
CI_RETRY_EXE=retry --
CI_IMAGE_LABEL=bitcoin-ci-test
BINS_SCRATCH_DIR=/ci_container_base/ci/scratch/bins/
TEST_RUNNER_TIMEOUT_FACTOR=40
PACKAGES=python3-zmq g++-arm-linux-gnueabihf busybox libc6:armhf libstdc++6:armhf libfontconfig1:armhf libxcb1:armhf
MAKEJOBS=-j4
CONTAINER_NAME=ci_arm_linux
PWD=/home/hebasto/git/bitcoin
CCACHE_DIR=/ci_container_base/ci/scratch/ccache
CI_BASE_PACKAGES=build-essential pkgconf curl ca-certificates ccache python3 rsync git procps bison e2fsprogs cmake
BASE_ROOT_DIR=/ci_container_base
DEPENDS_DIR=/ci_container_base/depends
RUN_FUNCTIONAL_TESTS=false
GOAL=install
CCACHE_COMPRESS=1
DPKG_ADD_ARCH=armhf
RUN_TIDY=false
HOST=arm-linux-gnueabihf
RUN_FUZZ_TESTS=false
CI_IMAGE_PLATFORM=linux/arm64
USE_BUSY_BOX=true
BOOST_TEST_RANDOM=1
BITCOIN_CONFIG=-DREDUCE_EXPORTS=ON -DCMAKE_CXX_FLAGS='-Wno-psabi -Wno-error=maybe-uninitialized'
SHLVL=2
DIR_QA_ASSETS=/ci_container_base/ci/scratch/qa-assets
CI_IMAGE_NAME_TAG=docker.io/debian:bookworm
BASE_SCRATCH_DIR=/ci_container_base/ci/scratch
BASE_READ_ONLY_DIR=/home/hebasto/git/bitcoin
LC_ALL=C.UTF-8
PREVIOUS_RELEASES_DIR=/ci_container_base/prev_releases
DEBIAN_FRONTEND=noninteractive
_=/usr/bin/python3
+ docker run --platform=linux/arm64 --rm docker.io/debian:bookworm bash -c 'env | grep --extended-regexp '\''^(HOME|PATH|USER)='\'''
+ tee --append /tmp/env-hebasto-ci_arm_linux
exec /usr/bin/bash: exec format error
+ MAYBE_CPUSET=
+ '[' '' ']'
+ echo 'Creating docker.io/debian:bookworm container to run in'
Creating docker.io/debian:bookworm container to run in
+ DOCKER_BUILDKIT=1
+ docker build --file /home/hebasto/git/bitcoin/ci/test_imagefile --build-arg CI_IMAGE_NAME_TAG=docker.io/debian:bookworm --build-arg FILE_ENV=./ci/test/00_setup_env_arm.sh --platform=linux/arm64 --label=bitcoin-ci-test --tag=ci_arm_linux /home/hebasto/git/bitcoin
[+] Building 0.1s (8/8) FINISHED                                                                                                                                                                                     docker:default
 => [internal] load build definition from test_imagefile                                                                                                                                                                       0.0s
 => => transferring dockerfile: 600B                                                                                                                                                                                           0.0s
 => WARN: InvalidDefaultArgInFrom: Default value for ARG ${CI_IMAGE_NAME_TAG} results in empty or invalid base image name (line 8)                                                                                             0.0s
 => [internal] load metadata for docker.io/library/debian:bookworm                                                                                                                                                             0.0s
 => [internal] load .dockerignore                                                                                                                                                                                              0.0s
 => => transferring context: 2B                                                                                                                                                                                                0.0s
 => [1/4] FROM docker.io/library/debian:bookworm                                                                                                                                                                               0.0s
 => [internal] load build context                                                                                                                                                                                              0.0s
 => => transferring context: 249B                                                                                                                                                                                              0.0s
 => CACHED [2/4] COPY ./ci/retry/retry /usr/bin/retry                                                                                                                                                                          0.0s
 => CACHED [3/4] COPY ./ci/test/00_setup_env.sh ././ci/test/00_setup_env_arm.sh ./ci/test/01_base_install.sh /ci_container_base/ci/test/                                                                                       0.0s
 => ERROR [4/4] RUN ["bash", "-c", "cd /ci_container_base/ && set -o errexit && source ./ci/test/00_setup_env.sh && ./ci/test/01_base_install.sh"]                                                                             0.1s
------                                                                                                                                                                                                                              
 > [4/4] RUN ["bash", "-c", "cd /ci_container_base/ && set -o errexit && source ./ci/test/00_setup_env.sh && ./ci/test/01_base_install.sh"]:
0.099 exec /usr/bin/bash: exec format error
------

 1 warning found (use docker --debug to expand):
 - InvalidDefaultArgInFrom: Default value for ARG ${CI_IMAGE_NAME_TAG} results in empty or invalid base image name (line 8)
test_imagefile:16
--------------------
  14 |     COPY ./ci/test/00_setup_env.sh ./${FILE_ENV} ./ci/test/01_base_install.sh /ci_container_base/ci/test/
  15 |     
  16 | >>> RUN ["bash", "-c", "cd /ci_container_base/ && set -o errexit && source ./ci/test/00_setup_env.sh && ./ci/test/01_base_install.sh"]
  17 |     
--------------------
ERROR: failed to solve: process "bash -c cd /ci_container_base/ && set -o errexit && source ./ci/test/00_setup_env.sh && ./ci/test/01_base_install.sh" did not complete successfully: exit code: 1
```

hebasto reviewed (approved):
ACK 6e29de21010fc5213176a6ba29f754ca72612ea0
